src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,55 +25,73 @@ class DictionarySegment : public BaseSegment {
-  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override;
+  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override {
+    const auto dictionary_value = _dictionary->at(_attribute_vector->at(chunk_offset));
+    return static_cast<AllTypeVariant>(dictionary_value);
+  }
 
   // return the value at a certain position.
-  T get(const size_t chunk_offset) const;
+  T get(const size_t chunk_offset) const { return _dictionary->at(_attribute_vector->at(chunk_offset)); }
 
   // dictionary segments are immutable
-  void append(const AllTypeVariant& val) override;
+  void append(const AllTypeVariant& val) override { throw std::runtime_error("Cannot append to immutable segments"); }
 
-  // returns an underlying dictionary
-  std::shared_ptr<const std::vector<T>> dictionary() const;
+  // returns the underlying dictionary
+  std::shared_ptr<const std::vector<T>> dictionary() const { return _dictionary; }
 
-  // returns an underlying data structure
-  std::shared_ptr<const BaseAttributeVector> attribute_vector() const;
+  // returns the attribute vector
+  std::shared_ptr<const BaseAttributeVector> attribute_vector() const { return _attribute_vector; }
 
   // return the value represented by a given ValueID
-  const T& value_by_value_id(ValueID value_id) const;
+  const T& value_by_value_id(ValueID value_id) { return _dictionary->at(value_id); }
 
   // returns the first value ID that refers to a value >= the search value
   // returns INVALID_VALUE_ID if all values are smaller than the search value
-  ValueID lower_bound(T value) const;
+  ValueID lower_bound(T value) const {
+    const auto lower_bound_it = std::lower_bound(_dictionary->begin(), _dictionary->end(), value);
+    if (lower_bound_it == _dictionary->end()) {
+      return INVALID_VALUE_ID;
+    }
+    return ValueID{std::distance(_dictionary->begin(), lower_bound_it)};
+  }
 
   // same as lower_bound(T), but accepts an AllTypeVariant
-  ValueID lower_bound(const AllTypeVariant& value) const;
+  ValueID lower_bound(const AllTypeVariant& value) const {
+    const T typed_value = static_cast<T>(value);
+    return lower_bound(typed_value);
+  }
 
   // returns the first value ID that refers to a value > the search value
   // returns INVALID_VALUE_ID if all values are smaller than or equal to the search value
-  ValueID upper_bound(T value) const;
+  ValueID upper_bound(T value) const {
+    const auto upper_bound_it = std::upper_bound(_dictionary->begin(), _dictionary->end(), value);
+    if (upper_bound_it == _dictionary->end()) {
+      return INVALID_VALUE_ID;
+    }
+    return ValueID{std::distance(_dictionary->begin(), upper_bound_it)};
+  }
 
   // same as upper_bound(T), but accepts an AllTypeVariant
-  ValueID upper_bound(const AllTypeVariant& value) const;
+  ValueID upper_bound(const AllTypeVariant& value) const {
+    const T typed_value = static_cast<T>(value);
```

Comment:
We used type_cast instead of static_cast but we are not sure which one is more correct.